### PR TITLE
CBO metric (number of dependencies)

### PR DESCRIPTION
**Pull request by simsimkic**
CBO metric counts the number of dependencies a class has. The calculation checks for any type used in the entire class:
- [x] field declaration
- [x] method return types
- [x] variable declarations

New model classes:
- CaDETVariable
- CaDETLinkedType

CaDETField is extended with:
- type

CaDETMember is extended with:
- return type
- list of variables

### CODE REVIEW

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs [hunk 1]
```diff
@@ -105,7 +106,8 @@ private List<CaDETField> ParseFields(IEnumerable<MemberDeclarationSyntax> nodeMe
                     {
                         Name = field.Identifier.Text,
                         Parent = parent,
-                        Modifiers = GetModifiers(node)
+                        Modifiers = GetModifiers(node),
+                        Type = new CaDETClass() { Name = fieldDeclaration.Declaration.Type.ToString()}
```
Comment (simsimkic) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs:110]: I used the Name attribute (not FullName), because when declaring fields, type is just _name_ of the type (not _namespace.name_).

**Comment by simsimkic** [comment]
[Kreirati fajlove za skup podataka](https://trello.com/c/sFKfFhfV/180-kreirati-fajlove-za-skup-podataka)

File: RepositoryCompiler/CodeModel/CaDETModel/CodeItems/CaDETClass.cs [hunk 2]
```diff
@@ -26,6 +26,9 @@ public string ContainerName
         public List<CaDETMember> Members { get; internal set; }
         public List<CaDETField> Fields { get; internal set; }
         public Dictionary<CaDETMetric, double> Metrics { get; internal set; }
+        public List<CaDETClass> FieldTypes { get; internal set; }
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CaDETModel/CodeItems/CaDETClass.cs:29]: This can be calculated fields (i.e., methods) instead of having duplicated data. E.g., GetFieldTypes returns Members.Fields.SelectMany(f => f.Type);

This saves memory and processing power, since we will only sometimes want to access a class's fieldTypes and when we do we will call this only once.

File: RepositoryCompiler/CodeModel/CaDETModel/CodeItems/CaDETMember.cs [hunk 3]
```diff
@@ -15,7 +15,9 @@ public class CaDETMember
         public ISet<CaDETMember> InvokedMethods { get; internal set; }
         public ISet<CaDETMember> AccessedAccessors { get; internal set; }
         public ISet<CaDETField> AccessedFields { get; internal set; }
+        public CaDETClass ReturnType { get; internal set; }
         public List<string> VariableNames { get; internal set; }
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CaDETModel/CodeItems/CaDETMember.cs:19]: VariableName and VariableType should be encapsulated into a CaDETVariable. Note that this will cause compilation errors in the BasicNamingStrategy of the SmartTutor so modify that code as well and run tests for both RC and ST.

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCaDETMemberBuilder.cs [hunk 4]
```diff
@@ -34,7 +34,9 @@ internal CaDETMember CreateBasicMember(CaDETClass parent)
             _member.SourceCode = _cSharpMember.ToString();
             _member.Parent = parent;
             _member.Params = GetMethodParams();
+            _member.ReturnType = GetMethodReturnType();
             _member.VariableNames = GetMethodVariableNames();
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCaDETMemberBuilder.cs:38]: GetMethodVariables to determine both.

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs (hunk 1 above)
Comment (Luburic) [reply RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs:110]: So long as you are sure this will uniquely identify the type. Did you consider situations where there are multiple classes with the same name (but in a different namespace)?

File: RepositoryCompilerTests/Unit/MetricCalculationTests.cs [hunk 5]
```diff
@@ -163,6 +163,22 @@ public void Calculates_max_nested_blocks_in_class()
             firstClass.Metrics[CaDETMetric.CMNB].ShouldBe(4);
         }
 
+        [Fact]
+        public void Calculates_number_of_dependencies_in_class()
```
Comment (Luburic) [inline RepositoryCompilerTests/Unit/MetricCalculationTests.cs:167]: Since you added fields, variables, and return types, you should expand the other TestSuite in this package to check if your types are correctly determined. Make sure your tests include classes with the same name.

At one point we will need to refactor RepoCompiler tests to have better test data (the factory has gotten out of hand) - e.g., a complete project that we can thoroughly test with all edge cases.

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CaDETClassMetricCalculator.cs [hunk 6]
```diff
@@ -188,5 +189,16 @@ private int CountUniqueMethodInvocations(CaDETClass parsedClass)
...
+
+        // Implementation based on https://github.com/mauricioaniche/ck
+        private int CountDependencies(CaDETClass parsedClass)
+        {
+            List<CaDETClass> allDependencies = new List<CaDETClass>();
+            allDependencies.AddRange(parsedClass.FieldTypes);
+            allDependencies.AddRange(parsedClass.MethodReturnTypes);
+            allDependencies.AddRange(parsedClass.MethodVariableTypes);
+            var uniqueDependencies = allDependencies.GroupBy(d => d.FullName).Select(d => d.First());
+            return uniqueDependencies.ToList().Count();
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CaDETClassMetricCalculator.cs:201]: Enumerable has a Count method.

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs [hunk 7]
```diff
@@ -199,6 +204,78 @@ private CaDETClass LinkOuterClass(List<CaDETClass> classes, string containerName
         {
             return classes.FirstOrDefault(c => c.FullName.Equals(containerName));
         }
+        
+        private List<CaDETClass> LinkFieldTypes(List<CaDETClass> classes, List<CaDETField> fields)
+        {
+            List<CaDETClass> fieldTypes = new List<CaDETClass>();
+            foreach (var f in fields)
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs:211]: Can this loop be replaced with fieldTypes.AddRange(FindTypes(classes, fields.Select(f => f.Type)))?

In light of the above comment I see that this approach is not suitable. We need to change each field to set a new Type for it.

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs [hunk 8]
```diff
@@ -199,6 +204,78 @@ private CaDETClass LinkOuterClass(List<CaDETClass> classes, string containerName
...
+            return returnTypes.Where(t => t != null).Distinct().ToList();
+        }
+
+        private List<CaDETClass> LinkMethodVariableTypes(List<CaDETClass> classes, List<CaDETMember> members)
+        {
+            var methods = members.Where(m => m.Type == CaDETMemberType.Method).ToList();
+            List<CaDETClass> variableTypes = new List<CaDETClass>();
+            foreach (var m in methods)
+            {
+                variableTypes.AddRange(FindTypes(classes, m.VariableTypes));
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs:234]: SelectMany

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs [hunk 9]
```diff
@@ -199,6 +204,78 @@ private CaDETClass LinkOuterClass(List<CaDETClass> classes, string containerName
...
+            foreach (var m in methods)
+            {
+                returnTypes.AddRange(FindTypes(classes, new List<CaDETClass>() { m.ReturnType}));
+            }
+            return returnTypes.Where(t => t != null).Distinct().ToList();
+        }
+
+        private List<CaDETClass> LinkMethodVariableTypes(List<CaDETClass> classes, List<CaDETMember> members)
+        {
+            var methods = members.Where(m => m.Type == CaDETMemberType.Method).ToList();
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs:230]: We are ignoring constructors?

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs [hunk 10]
```diff
@@ -182,6 +184,9 @@ private List<CaDETClass> ConnectCaDETGraph(List<CaDETClass> classes)
             {
                 c.Parent = LinkParent(classes, c.Parent);
                 c.OuterClass = LinkOuterClass(classes, c.ContainerName);
+                c.FieldTypes = LinkFieldTypes(classes, c.Fields);
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs:187]: You aren't really linking the types. You are generating the collections for the class, but the fields, variables, and return type of a method are not changed, meaning that our model is populated with dummy references that can't be used.

Once you remove these fields from the class the problem will present itself.

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs [hunk 11]
```diff
@@ -199,6 +204,78 @@ private CaDETClass LinkOuterClass(List<CaDETClass> classes, string containerName
...
+            }
+            return fieldTypes.Where(t => t != null).Distinct().ToList();
+        }
+        private List<CaDETClass> LinkMethodReturnTypes(List<CaDETClass> classes, List<CaDETMember> members)
+        {
+            var methods = members.Where(m => m.Type == CaDETMemberType.Method).ToList();
+            List<CaDETClass> returnTypes = new List<CaDETClass>();
+            foreach (var m in methods)
+            {
+                returnTypes.AddRange(FindTypes(classes, new List<CaDETClass>() { m.ReturnType}));
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs:223]: Same as above.

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs [hunk 12]
```diff
@@ -199,6 +204,78 @@ private CaDETClass LinkOuterClass(List<CaDETClass> classes, string containerName
...
+            }
+            return variableTypes.Where(t => t != null).Distinct().ToList();
+        }
+
+        private List<CaDETClass> FindTypes(List<CaDETClass> classes, List<CaDETClass> types)
+        {
+            List<CaDETClass> foundTypes = new List<CaDETClass>();
+            foreach (var type in types)
+            {
+                foundTypes.Add(classes.FirstOrDefault(c => c.Name.Equals(type.Name)));
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs:244]: This foreach will add a bunch of nulls, which is why you need to repeat the Where(t => t != null) in all of the above methods instead of doing it once at the end of this method.

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs [hunk 13]
```diff
@@ -199,6 +204,78 @@ private CaDETClass LinkOuterClass(List<CaDETClass> classes, string containerName
...
+        }
+
+        private List<CaDETClass> GetCollectionTypes(List<CaDETClass> classes, string type)
+        {
+            List<CaDETClass> collectionTypes = new List<CaDETClass>();
+            var match = Regex.Match(type, "<.+>");
+            if (!match.Success) return collectionTypes;
+
+            var value = match.Value.Substring(1, match.Value.Length - 2);
+            var types = value.Split(",").Select(t => t.Trim()).ToList();
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs:258]: Longer variable names would be beneficial to more clearly express the intent of each segment of the code.

"value" could be "typeParameterSection" using the terminology from https://docs.oracle.com/javase/tutorial/java/generics/types.html

"types" could be typeParameters

**Review by Luburic — CHANGES_REQUESTED** [review]
Comments inline.

File: RepositoryCompiler/CodeModel/CaDETModel/CodeItems/CaDETClass.cs (hunk 2 above)
Comment (simsimkic) [reply RepositoryCompiler/CodeModel/CaDETModel/CodeItems/CaDETClass.cs:29]: Member nema Fields?

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs (hunk 1 above)
Comment (simsimkic) [reply RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs]: Changed - full name is taken into account.

File: RepositoryCompiler/CodeModel/CaDETModel/CodeItems/CaDETLinkedType.cs [hunk 14]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+
+namespace RepositoryCompiler.CodeModel.CaDETModel.CodeItems
+{
+    public class CaDETLinkedType
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CaDETModel/CodeItems/CaDETLinkedType.cs:5]: A summary comment would be helpful here. It should contain a brief description of what the linkedType is and an example or two.

File: RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs [hunk 15]
```diff
@@ -88,24 +89,26 @@ private CaDETClass ParseClass(SemanticModel semanticModel, ClassDeclarationSynta
...
         {
             List<CaDETField> fields = new List<CaDETField>();
+
             foreach (var node in nodeMembers)
             {
                 if (!(node is FieldDeclarationSyntax fieldDeclaration)) continue;
+                
                 fields.AddRange(fieldDeclaration.Declaration.Variables.Select(
-                    field => new CaDETField
+                    variable => new CaDETField
```
Comment (Luburic) [inline RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpCodeParser.cs:106]: Why is this renamed to "variable"?

File: RepositoryCompilerTests/Unit/CaDETCodeModelTests.cs [hunk 16]
```diff
@@ -212,16 +212,80 @@ public void Calculates_method_variables()
...
+            List<CaDETClass> classes = factory.CreateClassModel(_testDataFactory.GetMultipleClassTexts());
+            List<CaDETClass> otherClasses = factory.CreateClassModel(_testDataFactory.GetClassesFromDifferentNamespace());
+
+            var dateRange = classes.Find(c => c.Name.Equals("DateRange"));
+            var doctor = classes.Find(c => c.Name.Equals("Doctor"));
+            var doctorService = classes.Find(c => c.Name.Equals("DoctorService"));
+            var otherDoctor = otherClasses.Find(c => c.Name.Equals("Doctor"));
+            var otherDateRange = otherClasses.Find(c => c.Name.Equals("DateRange"));
+
+            doctor.FindMember("TestFunction").Variables.SelectMany(v => v.GetLinkedTypes()).ShouldContain(doctor);
```
Comment (Luburic) [inline RepositoryCompilerTests/Unit/CaDETCodeModelTests.cs:284]: doctor.FindMember("TestFunction").Variables.SelectMany(v => v.GetLinkedTypes()) is duplicated 5 times. Change to:

var variables = doctor.FindMember("TestFunction").Variables.SelectMany(v => v.GetLinkedTypes());
variables.ShouldContain...

**Review by Luburic — COMMENTED** [review]
Several minor comments inline to address.

Make sure you run RepoCompiler unit and integration tests (for the projects we processed so far), as well as the SD, ST, and DE tests which shouldn't be affected, but still to be safe.

**Review by Luburic — APPROVED** [review]
Looks good.
